optbuilder: refactor scope and column_props

rytaft opened the pull request:
This commit contains a minor refactoring of the `scope.go` and
`column_props.go` files in optbuilder. The main changes are:
- `column_props.go` is renamed to `scope_column.go`, and the
  `columnProps` struct is therefore renamed to `scopeColumn`. This
  name better explains the function of the struct: it represents
  a bound variable in a particular scope created as part of the
  build process.
- Both `scopeColumn` and `scope` have a new `group` field, which
  contains the memo group ID of the scalar or relational expression
  in the current scope after it has been added to the memo.

These changes allow some amount of simplification in other parts
of the builder code. Most notably, buildXXX functions no longer
need to return `(out memo.GroupID, outScope *scope)`. They can simply
return `(outScope *scope)`, since the memo group ID is already
contained in `outScope.group`.

Release note: None

rytaft committed:
optbuilder: refactor scope and column_props

This commit contains a minor refactoring of the scope.go and
column_props.go files in optbuilder. The main changes are:
- column_props.go is renamed to scope_column.go, and the
  columnProps struct is therefore renamed to scopeColumn. This
  name better explains the function of the struct: it represents
  a bound variable in a particular scope created as part of the
  build process.
- Both scopeColumn and scope have a new group field, which
  contains the memo group ID of the scalar or relational expression
  in the current scope after it has been added to the memo.

These changes allow some amount of simplification in other parts
of the builder code. Most notably, buildXXX functions no longer
need to return (out memo.GroupID, outScope *scope). They can simply
return (outScope *scope), since the memo group ID is already
contained in outScope.group.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/24424)
<!-- Reviewable:end -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

Nice cleanup!

---

Review status: 0 of 15 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[pkg/sql/opt/optbuilder/builder.go, line 156 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L96mnFkBXXsN5DBmtLY:-L96mnFkBXXsN5DBmtLZ:bh3mqls) ([raw file](https://github.com/cockroachdb/cockroach/blob/be03a7213a5549ee208c96b20e791e3c787ed32b/pkg/sql/opt/optbuilder/builder.go#L156)):*
> ```Go
> //           "parent" scope that is still visible. The top-level memo group ID
> //           for the built statement/expression is returned in outScope.group.
> //           (Note that some buildXXX() functions pass outScope as an input
> ```

I don't understand this note.

---

*[pkg/sql/opt/optbuilder/groupby.go, line 252 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L96ngP_-WEnZ8bQmkle:-L96ngP_-WEnZ8bQmklf:bccrde4) ([raw file](https://github.com/cockroachdb/cockroach/blob/be03a7213a5549ee208c96b20e791e3c787ed32b/pkg/sql/opt/optbuilder/groupby.go#L252)):*
> ```Go
> //  - grouping columns
> //  - arguments to aggregate functions
> func (b *Builder) buildPreProjection(fromScope *scope, aggInScope *scope) *scope {
> ```

I think this should either return the memo group (and the caller sets the group), or return nothing (because `aggInScope = b.buildPreProjection(fromScope, aggInScope)` suggests we get a different scope back). In the latter case, the comment should mention that it modifies `aggInScope.group`

---

*[pkg/sql/opt/optbuilder/limit.go, line 28 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L96pqrq9NvLea-tG6dF:-L96pqrq9NvLea-tG6dG:b-gt19fr) ([raw file](https://github.com/cockroachdb/cockroach/blob/be03a7213a5549ee208c96b20e791e3c787ed32b/pkg/sql/opt/optbuilder/limit.go#L28)):*
> ```Go
> //   SELECT k FROM kv LIMIT k
> // are not valid.
> func (b *Builder) buildLimit(limit *tree.Limit, parentScope, inScope *scope) (outScope *scope) {
> ```

Same here, we always return `inScope`

---

*[pkg/sql/opt/optbuilder/select.go, line 238 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L96s7Cl-GZkYLUtp654:-L96s7Cl-GZkYLUtp655:bemtpt2) ([raw file](https://github.com/cockroachdb/cockroach/blob/be03a7213a5549ee208c96b20e791e3c787ed32b/pkg/sql/opt/optbuilder/select.go#L238)):*
> ```Go
> 	for _, table := range from.Tables {
> 		var rightScope *scope
> 		rightScope = b.buildTable(table, inScope)
> ```

rightScope :=

---

*[pkg/sql/opt/optbuilder/select.go, line 261 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L96sQgc3Q50cAradmME:-L96sQgc3Q50cAradmMF:br3uqrx) ([raw file](https://github.com/cockroachdb/cockroach/blob/be03a7213a5549ee208c96b20e791e3c787ed32b/pkg/sql/opt/optbuilder/select.go#L261)):*
> ```Go
> 		)
> 	} else {
> 		outScope.group = left
> ```

This is a no-op. Do we need `left` at all anymore? Seems like we could just do `if outScope == nil` instead.

---

*[pkg/sql/opt/optbuilder/util.go, line 107 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L96rjGw-qo2iiTtuAwF:-L96rjGw-qo2iiTtuAwG:bt7psx7) ([raw file](https://github.com/cockroachdb/cockroach/blob/be03a7213a5549ee208c96b20e791e3c787ed32b/pkg/sql/opt/optbuilder/util.go#L107)):*
> ```Go
> // typ    The type of the column.
> // expr   The expression this column refers to (if any).
> // group  The memo group ID of this column/expression.
> ```

[nit] Mention that this is optional (and can be set later in the returned scopeColumn).

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-:-L96pL619TlRUJg0pOPc:bjowcgf)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Thanks!

---

Review status: 0 of 15 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/sql/opt/optbuilder/builder.go, line 156 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L96mnFkBXXsN5DBmtLY:-L9756-I47FJvtOnzz7q:bsu3rcd) ([raw file](https://github.com/cockroachdb/cockroach/blob/be03a7213a5549ee208c96b20e791e3c787ed32b/pkg/sql/opt/optbuilder/builder.go#L156)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I don't understand this note.
</blockquote></details>

Moved it up to the note above. Hopefully that makes more sense?

---

*[pkg/sql/opt/optbuilder/groupby.go, line 252 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L96ngP_-WEnZ8bQmkle:-L975L_U-_hpMLn97iLs:b-6kgqkh) ([raw file](https://github.com/cockroachdb/cockroach/blob/be03a7213a5549ee208c96b20e791e3c787ed32b/pkg/sql/opt/optbuilder/groupby.go#L252)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I think this should either return the memo group (and the caller sets the group), or return nothing (because `aggInScope = b.buildPreProjection(fromScope, aggInScope)` suggests we get a different scope back). In the latter case, the comment should mention that it modifies `aggInScope.group`
</blockquote></details>

I removed this function and updated `constructProject` based on your suggestion.

---

*[pkg/sql/opt/optbuilder/limit.go, line 28 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L96pqrq9NvLea-tG6dF:-L975U5O0jcBt-Fzsz3K:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be03a7213a5549ee208c96b20e791e3c787ed32b/pkg/sql/opt/optbuilder/limit.go#L28)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Same here, we always return `inScope`
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/select.go, line 238 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L96s7Cl-GZkYLUtp654:-L975uwVAOzOQ2CrgISX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be03a7213a5549ee208c96b20e791e3c787ed32b/pkg/sql/opt/optbuilder/select.go#L238)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

rightScope :=
</blockquote></details>

Done.

---

*[pkg/sql/opt/optbuilder/select.go, line 261 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L96sQgc3Q50cAradmME:-L975vf3COpoiGWynCMy:b-4k0685) ([raw file](https://github.com/cockroachdb/cockroach/blob/be03a7213a5549ee208c96b20e791e3c787ed32b/pkg/sql/opt/optbuilder/select.go#L261)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This is a no-op. Do we need `left` at all anymore? Seems like we could just do `if outScope == nil` instead.
</blockquote></details>

Good point! Done.

---

*[pkg/sql/opt/optbuilder/util.go, line 107 at r1](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L96rjGw-qo2iiTtuAwF:-L976-Nh1BYIEoAjDZDs:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/be03a7213a5549ee208c96b20e791e3c787ed32b/pkg/sql/opt/optbuilder/util.go#L107)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] Mention that this is optional (and can be set later in the returned scopeColumn).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-:-L9760VCAnAsTwJgqOTQ:b-ucfknh)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 15 files reviewed at latest revision, all discussions resolved, all commit checks successful.

---

*[pkg/sql/opt/optbuilder/select.go, line 237 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L97HP7W5o90ZobnjQjM:-L97HP7W5o90ZobnjQjN:bqv5l9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/fef2d82bcac46b6cf1427b3793f89ff3a5f14e71/pkg/sql/opt/optbuilder/select.go#L237)):*
> ```Go
> func (b *Builder) buildFrom(from *tree.From, where *tree.Where, inScope *scope) (outScope *scope) {
> 	for _, table := range from.Tables {
> 		rightScope := b.buildTable(table, inScope)
> ```

[nit] "right" doesn't make much sense anymore. Maybe "tableScope"

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-:-L97HilKCmncA8fxCJh7:bnfp4nl)*
<!-- Sent from Reviewable.io -->

rytaft commented:
Review status: 0 of 15 files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/sql/opt/optbuilder/select.go, line 237 at r2](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-L97HP7W5o90ZobnjQjM:-L97IVN-0DaqBckojvnP:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fef2d82bcac46b6cf1427b3793f89ff3a5f14e71/pkg/sql/opt/optbuilder/select.go#L237)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] "right" doesn't make much sense anymore. Maybe "tableScope"
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Reviewed 10 of 17 files at r1, 5 of 6 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424)*
<!-- Sent from Reviewable.io -->

rytaft commented:
bors r+

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/cockroachdb/cockroach/24424#-:-L97N2Xw35tg1ctROGKq:b-ydi9ay)*
<!-- Sent from Reviewable.io -->
